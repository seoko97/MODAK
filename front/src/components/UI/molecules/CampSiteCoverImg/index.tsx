import React, { useCallback, useEffect, useMemo } from "react";
import { useDispatch } from "react-redux";
import styled, { css } from "styled-components";
import Router from "next/router";
import Link from "@atoms/Link";

import Title from "@atoms/Title";
import Button from "@atoms/Button";
import LookIcon from "@icons/LookIcon";
import CommentIcon from "@icons/CommentIcon";
import HeartIcon from "@icons/HeartIcon";
import useModal from "@hooks/useModal";
import ReviewForm from "@modals/ReviewForm";
import { ICamp } from "@type/reducers/camp";
import { AppDispatch, useAppSelector } from "@store/configureStore";
import { bookmark, unbookmark } from "@reducers/camp/action";
import { bookmarked, unBookmarked } from "@reducers/camps";
import { url } from "@apis/.";

const CampSiteCoverImg = () => {
  const dispatch: AppDispatch = useDispatch();
  const {
    singleCamp,
    bookmark: bmState,
    unbookmark: ubmState,
  } = useAppSelector((state) => state.camp);

  const { me } = useAppSelector((state) => state.user);

  useEffect(() => {
    if (!singleCamp) Router.push("/");
  }, []);

  const {
    _id,
    name,
    lineIntro,
    thema,
    environment,
    totalBookmark,
    photos,
    totalReview,
    views,
    bookmark: bmList,
  } = singleCamp as ICamp;

  const [isOpen, onOpen, onClose] = useModal();
  const bookMarkedUser = useMemo(() => {
    if (!me) return false;
    return bmList.includes(me._id);
  }, [me?._id, bmList]);

  const onClickBookMark = useCallback(async () => {
    if (me) {
      const res = await dispatch(bookmark(_id));
      dispatch(bookmarked(res));
    }
  }, [_id]);

  const onClickUnBookMark = useCallback(async () => {
    if (me) {
      const res = await dispatch(unbookmark(_id));
      dispatch(unBookmarked(res));
    }
  }, [_id, bookMarkedUser]);

  return (
    <>
      <CampCoverImgBox photos={photos}>
        <CampCoverImgContent>
          {/* 캠핑장 이름 */}
          <Title size={35}>{name}</Title>

          {/* 한 줄 소개 */}
          <p>{lineIntro}</p>

          {/* 아이콘/ 서브택스트 */}
          <IconContainer>
            <IconBox>
              <RowIcon>
                <LookIcon size={20} />
                <span>{views || 0}</span>
              </RowIcon>
              <RowIcon>
                <CommentIcon size={20} />
                <span>{totalReview || 0}</span>
              </RowIcon>
              <RowIcon>
                <HeartIcon size={16} />
                <span>{totalBookmark || 0}</span>
              </RowIcon>
            </IconBox>

            <ButtonBox>
              <Button onClick={onOpen} name="후기작성" />
              {isOpen && <ReviewForm onClick={onClose} camp={singleCamp as ICamp} />}
              <Button
                name={bookMarkedUser ? "북마크 제거" : "북마크 추가"}
                onClick={bookMarkedUser ? onClickUnBookMark : onClickBookMark}
                disabled={bmState.loading || ubmState.loading}
              />
            </ButtonBox>
          </IconContainer>

          {/* 테마 & 환경 */}
          <ThemaBox>{makeTagList(thema, "thema")}</ThemaBox>
          <ThemaBox>{makeTagList(environment, "environment")}</ThemaBox>
        </CampCoverImgContent>
      </CampCoverImgBox>
    </>
  );
};

function makeTagList(list: string[], type: string) {
  const lists = list.map((el, idx) => (
    <Link href={`/search?${type}=${el}`} key={idx}>
      <span key={idx}># {el}</span>
    </Link>
  ));
  return lists;
}

export default CampSiteCoverImg;

const CampCoverImgBox = styled.div<Pick<ICamp, "photos">>`
  position: relative;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  margin-bottom: 20px;
  background-image: url(${({ photos }) =>
    photos[0] ? `${url}/image/${photos[0]}` : "/tent.jpeg"});
  background-size: cover;
  overflow: hidden;

  &:before {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    content: "";
    display: block;
    background-color: rgba(0, 0, 0, 0.3);
  }
`;

const CampCoverImgContent = styled.div`
  width: 100%;
  max-width: 1600px;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: flex-start;
  gap: 10px;

  box-sizing: border-box;
  color: #fff;
  padding: 50px;
  z-index: 2;

  & > p {
    font-weight: bold;
    font-size: 20px;
    font-style: italic;
  }

  @media (max-width: ${({ theme }) => theme.BP.TABLET}) {
    & > h2 {
      font-size: 26px;
    }

    & > p {
      font-size: 16px;
    }
  }
`;

const ButtonBox = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 12px;

  & > button {
    background-color: royalblue;
    color: #fff;
    padding: 8px 12px;
    border-radius: 12px;
  }
  & > button:last-child {
    background-color: crimson;
  }

  @media (max-width: ${({ theme }) => theme.BP.TABLET}) {
    gap: 5px;

    & > button {
      padding: 6px 8px;
    }
  }
`;

const IconContainer = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-between;

  @media (max-width: ${({ theme }) => theme.BP.TABLET}) {
    flex-direction: column;
    align-items: flex-start;
    gap: 10px;
  }
`;

const IconBox = styled.div`
  display: flex;
  gap: 15px;
`;

const CampSiteIcon = css`
  display: flex;
  align-items: center;
  gap: 5px;
  color: #fff;

  svg {
    fill: #fff;

    & + span {
      font-size: 14px;
    }
  }

  @media (max-width: ${({ theme }) => theme.BP.TABLET}) {
    svg {
      width: 15px;
      height: 15px;

      & + span {
        font-size: 12px;
      }
    }
  }
`;

const RowIcon = styled.div`
  ${CampSiteIcon};
`;

const ThemaBox = styled.div`
  display: flex;
  flex-wrap: wrap;
  gap: 20px;
  & > a {
    font-weight: bold;
    color: #fff;
    background-color: #038c5a;
    padding: 8px 10px;
    border-radius: 12px;

    @media (max-width: ${({ theme }) => theme.BP.TABLET}) {
      padding: 6px 8px;
    }
  }
`;
