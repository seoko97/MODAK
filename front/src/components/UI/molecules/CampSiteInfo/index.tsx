import React from "react";
import styled from "styled-components";
import { ICamp } from "@type/reducers/camp";
import { useAppSelector } from "@store/configureStore";

const CampSiteInfo = () => {
  const { singleCamp } = useAppSelector((state) => state.camp);

  const { address, intro, tel, animal, category, rental, firstImage } = singleCamp as ICamp;
  return (
    <>
      <CampSiteContainer>
        <CampSiteImage>
          <img alt="캠핑장 소개 사진" src={firstImage || "/tent.jpeg"} />
        </CampSiteImage>

        <CampSiteContentBox>
          {/* 소개글 */}
          <p>{intro}</p>

          {/* 상세정보(주소, 전화번호) */}
          <CampSiteDetailBox>
            <CampSiteDetail>
              <span>야영장 구분</span>
              <p>{makeList(category)}</p>
            </CampSiteDetail>
            <CampSiteDetail>
              <span>연락처</span>
              <p>{tel}</p>
            </CampSiteDetail>
            <CampSiteDetail>
              <span>주소</span>
              <p>{address}</p>
            </CampSiteDetail>
            <CampSiteDetail>
              <span>대여</span>
              <p>{makeList(rental)}</p>
            </CampSiteDetail>
            <CampSiteDetail>
              <span>반려동물출입</span>
              <p>{animal}</p>
            </CampSiteDetail>
          </CampSiteDetailBox>
        </CampSiteContentBox>
      </CampSiteContainer>
    </>
  );
};

function makeList(list: string[]) {
  const lists = list.map((el, idx) => <span key={idx}>{el} </span>);
  return lists;
}

export default CampSiteInfo;

// 캠핑장 소개 박스 (이미지 & 주소 등)
const CampSiteContainer = styled.div`
  position: relative;
  width: 100%;
  display: flex;
  box-sizing: border-box;
  align-items: flex-start;
  justify-content: center;
  & * {
    color: ${({ theme }) => theme.FONT_COLOR.PRIMARY_COLOR};
  }

  @media (max-width: ${({ theme }) => theme.BP.TABLET}) {
    flex-direction: column;
  }
`;

const CampSiteImage = styled.div`
  position: relative;
  width: 50%;
  min-height: 400px;

  img {
    position: absolute;
    width: 100%;
    height: 100%;
    object-fit: cover;
  }

  @media (max-width: ${({ theme }) => theme.BP.TABLET}) {
    width: 100%;
    height: 400px;
  }
`;

const CampSiteContentBox = styled.div`
  flex: 1;
  display: flex;
  flex-direction: column;
  box-sizing: border-box;
  padding: 5px 25px;
  gap: 15px;

  & > p {
    font-weight: 500;
  }

  @media (max-width: ${({ theme }) => theme.BP.TABLET}) {
    padding: 25px 5px;
  }
`;

const CampSiteDetailBox = styled.div`
  display: flex;
  flex-direction: column;
  gap: 5px;

  margin-top: auto;
`;

const CampSiteDetail = styled.div`
  display: flex;
  justify-content: flex-start;

  & > span:first-of-type {
    flex: 1;
    min-width: 100px;
    font-weight: 700;
  }

  & > p {
    flex: 4;
  }
`;
