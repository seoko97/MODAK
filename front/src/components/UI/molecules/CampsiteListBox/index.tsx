import React from "react";
import styled from "styled-components";
import Link from "@atoms/Link";

import BookmarkIcon from "@icons/BookmarkIcon";
import CommentIcon from "@icons/CommentIcon";
import { ICamp } from "@type/reducers/camp";

interface Props {
  camp: ICamp;
}

const CampSiteListBox = ({ camp }: Props) => {
  const { _id, name, address, tel, totalBookmark, totalReview, environment, thema, firstImage } =
    camp;
  return (
    <Link href={`/camp/${_id}`}>
      <CardWrapper>
        <ImgWrapper>
          <img src={firstImage || "/tent.jpg"} alt={`${name} 사진`} />
        </ImgWrapper>
        <CampSiteInfo>
          <CardInfoHeader>
            <h3>{name}</h3>
            <CountsWrapper>
              <CountContainer>
                <BookmarkIcon size={20} />
                <span>{totalBookmark || 0}</span>
              </CountContainer>
              <CountContainer>
                <CommentIcon size={20} />
                <span>{totalReview || 0}</span>
              </CountContainer>
            </CountsWrapper>
          </CardInfoHeader>
          <InfoTable>
            <tbody>
              <tr>
                <th>주소</th>
                <td>{address}</td>
              </tr>
              <tr>
                <th>연락처</th>
                <td>{tel}</td>
              </tr>
            </tbody>
          </InfoTable>
          <TagsContainer>
            {environment.map((tag) => (
              <li key={tag}>{tag}</li>
            ))}
            {thema.map((tag) => (
              <li key={tag}>{tag}</li>
            ))}
          </TagsContainer>
        </CampSiteInfo>
      </CardWrapper>
    </Link>
  );
};

CampSiteListBox.defaultProps = {
  photos: [
    "https://images.unsplash.com/photo-1607908560428-36ff9e0363b7?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1635&q=80",
  ],
  totalBookmark: 0,
  totalReview: 0,
};

export default CampSiteListBox;

const CardWrapper = styled.div`
  width: 100%;
  display: flex;
  margin: 1em auto 0;
  padding-bottom: 1em;
  border-bottom: 1px dotted #c0c0c0;
  color: ${({ theme }) => theme.FONT_COLOR.PRIMARY_COLOR};
  transition: 0.3s transform;
  font-size: 20px;

  &:hover {
    transform: scale(1.005);
    cursor: pointer;
  }

  @media (max-width: ${({ theme }) => theme.BP.TABLET}) {
    flex-direction: column;
    height: auto;
    padding-bottom: 0;
  }
  @media (max-width: ${({ theme }) => theme.BP.MOBILE}) {
    font-size: 1rem;
  }
`;

const ImgWrapper = styled.div`
  width: 40%;
  margin-right: 1em;
  position: relative;
  box-shadow: 0 2px 6px 0 hsla(0, 0%, 0%, 0.2);

  &:before {
    content: "";
    display: block;
    padding-top: 100%;
  }

  & > img {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    object-fit: cover;
  }

  @media screen and (max-width: ${({ theme }) => theme.BP.TABLET}) {
    width: auto;
    height: auto;
    margin-right: 0;
  }
`;

const CampSiteInfo = styled.div`
  width: 100%;
  height: 100%;
  position: relative;
  padding: 1em;
  font-size: 0.8em;
  display: flex;
  flex-direction: column;
  gap: 10px;
`;

const CardInfoHeader = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 10px;

  h3 {
    font-size: 1.2em;
    font-weight: 700;
  }
`;

const InfoTable = styled.table`
  border-collapse: separate;
  border-spacing: 0 5px;

  & th {
    text-align: left;
    width: 30%;
  }

  & td {
    padding-left: 5px;
  }
`;

const TagsContainer = styled.ul`
  display: flex;
  flex-wrap: wrap;

  & > li {
    margin-right: 15px;

    &::before {
      content: "#";
    }
  }
`;

const CountsWrapper = styled.div`
  display: flex;
`;

const CountContainer = styled.div`
  display: flex;

  svg {
    margin-right: 2px;
    margin-left: 10px;
  }
  @media (max-width: ${({ theme }) => theme.BP.TABLET}) {
    svg {
      margin-top: 2px;
      width: 15px;
      height: 15px;
    }
  }
`;
