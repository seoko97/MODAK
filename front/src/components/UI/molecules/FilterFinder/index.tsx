import React, { useState, useCallback, useEffect } from "react";
import { useDispatch } from "react-redux";
import styled from "styled-components";
import { getCamps } from "@reducers/camps/action";
import useScroll from "@hooks/useScroll";
import useThrottle from "@hooks/useThrottle";
import { AppDispatch, useAppSelector } from "@store/configureStore";
import FilterCategory from "./FilterCategory";

interface QueryProps {
  [key: string]: string[];
}

interface Props {
  sorted: string;
}

const categories = [
  {
    name: "지역",
    options: [
      "전체",
      "서울",
      "부산",
      "대구",
      "인천",
      "광주",
      "대전",
      "울산",
      "세종",
      "경기",
      "강원",
    ],
  },
  {
    name: "주변환경",
    options: ["해변", "산", "숲", "계곡", "강", "호수", "도심"],
  },
  {
    name: "부대시설",
    options: ["전기", "wifi", "온수", "수영장", "산책로", "편의점"],
  },
  {
    name: "테마",
    options: ["낚시", "일출명소", "일몰명소", "물놀이", "액티비티"],
  },
];

const FilterFinder = ({ sorted }: Props) => {
  const [query, setQuery] = useState<QueryProps>({});
  const [skip, setSkip] = useState<number>(0);
  const { mainCamps } = useAppSelector((state) => state.camps);
  const [scrollHeight, clientHeight] = useScroll();
  const dispatch: AppDispatch = useDispatch();

  const onThrottle = useThrottle(async () => {
    await dispatch(getCamps({ ...query, sorted, skip: `${skip + 1}` }));
    setSkip(skip + 1);
  }, 500);

  const checked = useCallback(
    (title: string, list: string[]) => {
      const newQuery = {
        ...query,
        [title]: list,
      };

      if (list[0] === "전체") newQuery[title] = [];
      setQuery(newQuery);
    },
    [query],
  );

  const searchCamps = useCallback(async () => {
    await dispatch(getCamps({ ...query, sorted }));
  }, [query, sorted]);

  useEffect(() => {
    if (scrollHeight + 300 >= clientHeight) {
      onThrottle();
    }
  }, [scrollHeight, clientHeight, query, mainCamps, sorted, skip]);

  return (
    <FinderContainer>
      {categories.map((category) => (
        <FilterCategory key={category.name} category={category} query={query} checked={checked} />
      ))}
      <ButtonContainer>
        <input type="button" value="검색" onClick={searchCamps} />
      </ButtonContainer>
    </FinderContainer>
  );
};

export default FilterFinder;

const FinderContainer = styled.form`
  width: 100%;
  margin-top: 1em;
  border: 1px solid #ebebeb;
  border-radius: 10px;
  box-shadow: 3px 4px 5px rgba(0, 0, 0, 0.1);
  padding: 30px 50px;
  color: ${({ theme }) => theme.FONT_COLOR.PRIMARY_COLOR};

  @media (max-width: ${({ theme }) => theme.BP.TABLET}) {
    padding: 20px;
  }
`;

const ButtonContainer = styled.div`
  display: flex;
  justify-content: flex-end;

  input {
    border-radius: 4px;
    border: none;
    padding: 10px 20px;

    &:hover {
      cursor: pointer;
    }
  }

  input + input {
    margin-left: 20px;
  }

  @media (max-width: ${({ theme }) => theme.BP.TABLET}) {
    input {
      width: 50px;
      height: 20px;
      font-size: 10px;
      padding: 0;
    }

    input + input {
      margin-left: 10px;
    }
  }
`;
