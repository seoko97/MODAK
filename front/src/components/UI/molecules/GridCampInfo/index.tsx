import Image from "next/image";
import Link from "next/link";
import React from "react";
import styled from "styled-components";
import GridCampInfoInner from "./GridCampInfoInner";

interface Props {
  to: string;
  title: string;
  description: string;
  src: string;
}

const StyledCamp = styled.div<Pick<Props, "src">>`
  width: 100%;
  height: 200px;
  position: relative;
  cursor: pointer;

  @media (max-width: ${({ theme }) => theme.BP.TABLET}) {
    height: 150px;
  }
`;

const GridCampInfo = ({ to, title, description, src }: Props) => {
  return (
    <>
      <Link href={to}>
        <StyledCamp src={src}>
          <Image alt="asd" src={src} layout="fill" objectFit="cover" />
          <GridCampInfoInner title={title} description={description} />
        </StyledCamp>
      </Link>
    </>
  );
};

export default GridCampInfo;
