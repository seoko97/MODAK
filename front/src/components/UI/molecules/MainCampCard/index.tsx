import React, { useMemo } from "react";
import styled from "styled-components";
import Link from "@atoms/Link";

import PinIcon from "@icons/Pin";
import BookmarkIcon from "@icons/BookmarkIcon";
import CommentIcon from "@icons/CommentIcon";
import { ICamp } from "@type/reducers/camp";

interface Props {
  camp: ICamp;
  url: string;
}

const Container = styled.div`
  width: 95%;
  cursor: pointer;
  margin: 0 auto;
  border-radius: 8px;
  box-shadow: 1px 2px 6px 0px rgb(158 154 153 / 32%);
  & * {
    color: ${({ theme }) => theme.FONT_COLOR.PRIMARY_COLOR};
  }
`;

const ImageWrapper = styled.div`
  width: 100%;

  & > div {
    position: relative;
    width: 100%;
    padding-bottom: 200px;
  }

  & img {
    border-radius: 8px 8px 0 0;
    position: absolute;
    width: 100%;
    height: 100%;
    object-fit: cover;
  }

  @media (max-width: ${({ theme }) => theme.BP.PC}) {
    & > div {
      padding-bottom: 70%;
      max-height: 320px;
    }
  }
`;

const ContentWrapper = styled.div`
  width: 100%;
  padding: 14px;
  display: flex;
  flex-direction: column;
  border-radius: 0 0 8px 8px;
`;

const ContentHeader = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  padding-bottom: 3px;
  gap: 8px;
  & > h1 {
    font-size: 18px;
    overflow: hidden;
  }
`;

const LocationWrapper = styled.div`
  display: flex;
  font-size: 12px;
  line-height: 12px;
  gap: 5px;
  margin-bottom: 5px;
  & > h2 {
    color: #757575;
  }
`;

const ThemaWrapper = styled.div`
  width: 100%;
  padding: 12px 0;
  display: flex;
  flex-wrap: wrap;
  gap: 10px;
  min-height: 60px;
  font-size: 12px;
`;

const ContentFooter = styled.div`
  width: 100%;
  display: flex;
  gap: 10px;
`;

const IconWrapper = styled.div`
  display: flex;
  gap: 5px;
  align-items: center;
  justify-content: center;
`;

const MainCampCard = ({ camp, url }: Props) => {
  const { name, address, thema, category, totalBookmark, views, firstImage } = camp;

  const tagList = useMemo(() => {
    const set = new Set([...thema, ...category]);
    const list: JSX.Element[] = [];
    set.forEach((e) => list.push(<span key={e}>#{e}</span>));
    return list;
  }, [thema, category]);

  return (
    <Link href={url}>
      <Container>
        <ImageWrapper>
          <div>
            <img src={firstImage || "/tent.jpg"} alt="img" />
          </div>
        </ImageWrapper>
        <ContentWrapper>
          <ContentHeader>
            <h1>{name}</h1>
            <LocationWrapper>
              <PinIcon size={16} />
              <h2>{address}</h2>
            </LocationWrapper>
          </ContentHeader>
          <ThemaWrapper>{tagList}</ThemaWrapper>
          <ContentFooter>
            <IconWrapper>
              <BookmarkIcon size={16} />
              <span>{totalBookmark || 0}</span>
            </IconWrapper>
            <IconWrapper>
              <CommentIcon size={16} />
              <span>{views || 0}</span>
            </IconWrapper>
          </ContentFooter>
        </ContentWrapper>
      </Container>
    </Link>
  );
};

export default MainCampCard;
