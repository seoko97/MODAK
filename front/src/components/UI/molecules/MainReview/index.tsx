import Link from "@atoms/Link";
import React from "react";
import styled from "styled-components";
import { IReview } from "@type/reducers/review";
import { url } from "@apis/.";
import Avatar from "@atoms/Avatar";

import LocationIcon from "@icons/LocationIcon";
import CalendarIcon from "@icons/CalendarIcon";
import HeartIcon from "@icons/HeartIcon";
import { checkUrl } from "@lib/checkUrl";

interface Props {
  review: IReview;
}
const limitStr = (str: string) => (str.length >= 25 ? `${str.substring(0, 25)}...` : str);

const MainReview = ({ review }: Props) => {
  const { author, content, count, photos, created, location } = review;
  const { nickname, profileImg } = author;

  return (
    <>
      <Link href={`/camp/${location._id}`}>
        <StyledMainReviewContainer>
          <StyledCardImage>
            <div>
              <img alt="Mountains" src={photos[0] ? `${url}/image/${photos[0]}` : "/tent.jpeg"} />
            </div>
          </StyledCardImage>

          {/* 카드내용 : 유저정보, 리뷰정보로 구성 */}
          <StyledCardContentBox>
            <StyledUserInfo>
              <Avatar url={checkUrl(profileImg)} alt="사진" />
              <span>{nickname}</span>
            </StyledUserInfo>

            <StyledReviewInfo>
              <StyledReviewIconBox>
                <LocationIcon size={20} />
                <span>{location.name}</span>
              </StyledReviewIconBox>

              <div>
                <p>{limitStr(content)}</p>
              </div>

              <StyledDateLike>
                <StyledReviewIconBox>
                  <CalendarIcon size={20} />
                  <span>{created}</span>
                </StyledReviewIconBox>

                <StyledReviewIconBox>
                  <HeartIcon size={20} />
                  <span>{count.toLocaleString()}</span>
                </StyledReviewIconBox>
              </StyledDateLike>
            </StyledReviewInfo>
          </StyledCardContentBox>
        </StyledMainReviewContainer>
      </Link>
    </>
  );
};

export default MainReview;

// 리뷰 컨테이너
const StyledMainReviewContainer = styled.div`
  width: 95%;
  margin: 0 auto;
  border-radius: 8px;
  box-shadow: 1px 2px 6px 0px rgb(158 154 153 / 32%);

  & * {
    color: ${({ theme }) => theme.FONT_COLOR.PRIMARY_COLOR};
  }

  cursor: pointer;
`;

const StyledCardImage = styled.div`
  width: 100%;

  & > div {
    position: relative;
    width: 100%;
    padding-bottom: 200px;
  }

  & img {
    border-radius: 8px 8px 0 0;
    position: absolute;
    width: 100%;
    height: 100%;
    object-fit: cover;
  }

  @media (max-width: ${({ theme }) => theme.BP.PC}) {
    & > div {
      padding-bottom: 70%;
      max-height: 320px;
    }
  }
`;

const StyledCardContentBox = styled.div`
  width: 100%;
  height: 50%;
  display: flex;
  flex-direction: column;
  padding: 20px;
  gap: 10px;

  & > div:first-of-type {
    margin-bottom: 10px;
  }
`;

const StyledUserInfo = styled.div`
  display: flex;
  justify-content: flex-start;
  align-items: center;
  gap: 10px;
  font-weight: bold;
`;

const StyledReviewInfo = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  align-items: flex-start;
  gap: 20px;
`;

const StyledDateLike = styled.div`
  display: flex;
  justify-content: space-between;
  width: 100%;
  margin-top: auto;
`;

const StyledReviewIconBox = styled.div`
  display: flex;
  width: auto;
  align-items: center;

  gap: 5px;
  color: #757575;

  & svg {
    fill: #757575;
  }

  & > span {
    line-height: 1;
    font-size: 14px;
  }
`;
