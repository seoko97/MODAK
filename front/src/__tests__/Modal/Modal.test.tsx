import React from "react";
import UserUpdate from "@src/components/modals/UserUpdateForm";
import { User } from "@src/components/UI/molecules/MypageProfile";
import renderer from "react-test-renderer";
import { screen } from "@testing-library/react";

describe("Modal 테스트", () => {
  let onClick: () => void;
  let user: User;
  let updateUser: (userInfo: User) => void;
  let portalRoot: HTMLCollectionOf<Element>;
  beforeEach(() => {
    user = {
      _id: 1,
      firstName: "이",
      lastName: "태현",
      email: "asdasd@gmail.com",
      nickname: "현",
      profile:
        "https://images.unsplash.com/photo-1560250097-0b93528c311a?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=687&q=80",
      createAt: "1121-11-11",
      source: "source",
      likes: 1,
      intro:
        "Lorem ipsum dolor sit amet consectetur adipisicing elit. Illo eveniet magnam dolor quasi laboriosam, qui facere deleniti, assumenda eius, expedita quisquam sit. Vitae veniam corrupti voluptatum laboriosam culpa, accusantium quia!",
    };
    updateUser = jest.fn();
    onClick = jest.fn();
  });

  it("snapshot", () => {
    const component = renderer.create(
      <UserUpdate onClick={onClick} user={user} updateUser={updateUser} />,
    );
    // portalRoot = document.getElementById("modal");
    expect(component.toJSON()).toMatchSnapshot();
  });
});

// ? 실제 -> Mock -> Stub, spy..
// ? 하나의 오브젝트에 여러 function, 일부분만 fake로 만드
// ? mock보다 stub이 일부에 집중할 때
// ? spy : 진짜 오브젝트에 대해서 안쪽에 들어가게 해서, 거기서 수행하는 것들을 갈아끼우거나, 리턴하게 하는. 좀 하이브리드 느낌.
// ? 테스트 작성이 어려우면? 리얼 코드를 고쳐야하는 건 아닌지 생각해보자.
// ? Mock : 몇 번 호출됐는지 확인 가능한데, 스텁은 그런 게 좀 안 되는. (toHaveCalledTimes, toHaveCalledWith 등등...)

// ? toHaveCalledTimes, toHaveCalledWith: 하는 이유는?
// ? 몽구스 커넥션에 대해서, 호출을 잘 했는지 확인(1번만 호출됐느냐?), 내가 원하는 결과와 같이 왔느냐?()
// ? API 테스트를 할 때, 파라미터들이 같이 올 텐데, 쿼리문에 잘 포함되고 있는지..
// ? 타임 라이브러리에 A란 메소드를 잘 가져왔는지, 원하는 format으로 잘 가져왔는지...

// ? E2E(end to end) 테스트...보단 unit test를.
// ? 몽구스 자체보다는, 내가 만든 메소드가 잘 기능하는지.
// ? JSON이 string으로 왔으면, 그걸 다시 JSON으로 변환시켜서 테스트.
// ? 테스트를 문서화하면, 걔네들에 대한 스펙을 정리.
// ? get이라는 API에 대해서 이러이러하게 정리를 하고서, 테스트를 작성.
// ? 테스트를 먼저 만들고, 기능 개발을 구현.
// ? 테스트는 이 사람이, 기능은 저 사람이.

// ? 서버 개발을 하다보면, 리얼 테스트를 안 해도 될 때가 많음.
// ? 어떤 데이터가 들어갔는지, 테스트 코드가 있으면 안 해도 됨. 포스트맨에서 설정하고, 뭐하고, 파일 첨부하고, 키도 바꾸고...서버 저장 잘 됐는지...
// ? 테스트 코드 작성했으면? 쿼리가 잘 돌아가는지, 세이브가 해당 공간에 잘 되는지 확인 가능.

// ? 시간 절약이 매우 좋음.
// ? 테스트는 리팩토링에 굉장히 좋음. 조기에 발견하기 좋음.
// ? 구직에 도움 되는.

// ? TDD를 할 때 주니어와 시니어의 조화.
// ? 리액트만 보더라도, 빌드 시간 절약.

// ? 내가 만든 스펙을 미리 정리. 더 빠르게, 더 명확하게 개발.
// ? 구조를 잡기 좋음.
