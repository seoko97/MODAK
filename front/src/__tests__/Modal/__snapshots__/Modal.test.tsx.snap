// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal 테스트 snapshot 1`] = `null`;
